tcg/src/log.rs

```diff
@@ -0,0 +1,377 @@
+// Copyright (c) 2022 Intel Corporation
+//
+// SPDX-License-Identifier: BSD-2-Clause-Patent
+
+use super::*;
+use alloc::boxed::Box;
+use core::mem::size_of;
+
+#[derive(Debug)]
+pub enum CcEventLogError {
+    InvalidParameter,
+    OutOfResource,
+}
+
+#[allow(unused)]
+pub struct CcEventLogWritter<'a> {
+    area: &'a mut [u8],
+    offset: usize,
+    last: usize,
+    extender: Box<dyn Fn(&[u8; SHA384_DIGEST_SIZE], u32)>,
+}
+
+impl<'a> CcEventLogWritter<'a> {
+    pub fn new(
+        area: &mut [u8],
+        extender: Box<dyn Fn(&[u8; SHA384_DIGEST_SIZE], u32)>,
+    ) -> Option<CcEventLogWritter> {
+        let mut td_event_log = CcEventLogWritter {
+            area,
+            offset: 0,
+            last: 0,
+            extender,
+        };
+
+        // Create the TCG_EfiSpecIDEvent as the first event
+        let first = TcgEfiSpecIdevent::default();
+        td_event_log.log_pcr_event(0, EV_NO_ACTION, first.as_bytes())?;
+
+        Some(td_event_log)
+    }
+
+    pub fn create_event_log(
+        &mut self,
+        mr_index: u32,
+        event_type: u32,
+        event_data: &[&[u8]],
+        hash_data: &[u8],
+    ) -> Result<(), CcEventLogError> {
+        let sha384 = self.calculate_digest_and_extend(hash_data, mr_index);
+
+        self.log_cc_event(mr_index, event_type, event_data, &sha384)
+    }
+
+    pub fn create_seperator(&mut self) -> Result<(), CcEventLogError> {
+        let separator = u32::to_le_bytes(0);
+
+        // Measure 0x0000_0000 into RTMR[0] and RTMR[1]
+        let _ = self.calculate_digest_and_extend(&separator, 1);
+        let sha384 = self.calculate_digest_and_extend(&separator, 2);
+
+        self.log_cc_event(1, EV_SEPARATOR, &[&separator], &sha384)?;
+        self.log_cc_event(2, EV_SEPARATOR, &[&separator], &sha384)
+    }
+
+    pub fn as_slice(&self) -> &[u8] {
+        &self.area[..self.offset]
+    }
+
+    fn calculate_digest_and_extend(
+        &self,
+        hash_data: &[u8],
+        mr_index: u32,
+    ) -> [u8; SHA384_DIGEST_SIZE] {
+        let hash_value = ring::digest::digest(&ring::digest::SHA384, hash_data);
+        let hash_value = hash_value.as_ref();
+        assert_eq!(hash_value.len(), SHA384_DIGEST_SIZE);
```

Comment:
Just querying this assert. Would it be better to protect it with `#[cfg(test)]`, or should this be a `panic!()`, or should the fucntion return a `Result<>`?

Reply:
It should always be `SHA384_DIGEST_SIZE` when we use `SHA384` to calculate the digest, is it good to use assert here?